plugin/src/main/java/com/denizenscript/denizen/scripts/commands/entity/InvisibleCommand.java

```diff
@@ -73,87 +96,130 @@ else if (!scriptEntry.hasObject("target")
+
+    public static void setInvisibleForPlayer(EntityTag target, PlayerTag player, boolean invisible) {
+        if (target == null || target.getUUID() == null || player == null) {
+            return;
         }
-        if (!scriptEntry.hasObject("target") || !((EntityTag) scriptEntry.getObjectTag("target")).isValid()) {
-            throw new InvalidArgumentsException("Must specify a valid target!");
+        NetworkInterceptHelper.enable();
+        boolean wasModified = !invisibleEntities.containsKey(target.getUUID());
+        HashMap<UUID, Boolean> playerMap = invisibleEntities.computeIfAbsent(target.getUUID(), k -> new HashMap<>());
+        wasModified = !playerMap.containsKey(player.getUUID()) || playerMap.get(player.getUUID()) != invisible || wasModified;
+        playerMap.put(player.getUUID(), invisible);
+        if (wasModified) {
+            NMSHandler.packetHelper.sendEntityMetadataFlagsUpdate(player.getPlayerEntity(), target.getBukkitEntity());
         }
     }
 
-    public void setInvisible(EntityTag entity, boolean visible) {
-        if (entity.getBukkitEntity() instanceof ArmorStand) {
-            ((ArmorStand) entity.getBukkitEntity()).setVisible(visible);
+    public void setInvisible(EntityTag entity, boolean invisible) {
+        if (entity.isCitizensNPC()) {
+            entity.getDenizenNPC().getCitizen().getOrAddTrait(InvisibleTrait.class).setInvisible(invisible);
+        }
+        else if (entity.getBukkitEntity() instanceof ArmorStand) {
+            ((ArmorStand) entity.getBukkitEntity()).setVisible(!invisible);
         }
         else if (entity.getBukkitEntity() instanceof ItemFrame) {
-            ((ItemFrame) entity.getBukkitEntity()).setVisible(visible);
+            ((ItemFrame) entity.getBukkitEntity()).setVisible(!invisible);
         }
         else if (entity.isLivingEntity() && !entity.isFake) {
-            if (visible) {
+            entity.getLivingEntity().setInvisible(invisible);
+            if (!invisible) {
+                // Remove the invisibility potion effect for compact with old uses (the command used to add it)
                 entity.getLivingEntity().removePotionEffect(PotionEffectType.INVISIBILITY);
             }
-            else {
-                new PotionEffect(PotionEffectType.INVISIBILITY, Integer.MAX_VALUE, 1).apply(entity.getLivingEntity());
-            }
         }
         else {
-            NMSHandler.entityHelper.setInvisible(entity.getBukkitEntity(), !visible);
+            NMSHandler.entityHelper.setInvisible(entity.getBukkitEntity(), invisible);
         }
     }
 
     @Override
     public void execute(ScriptEntry scriptEntry) {
-        ElementTag state = scriptEntry.getElement("state");
         EntityTag target = scriptEntry.getObjectTag("target");
-        if (scriptEntry.dbCallShouldDebug()) {
-            Debug.report(scriptEntry, getName(), state, target);
+        if (target == null || target.isFake) {

```

Comment:
Why are you excluding fake entities here?

Reply:
I... think it did that somewhere else and I moved it here? might've confused it with something - removed

Reply:
The spot you're referring to isn't an exclusion, it's just a separate codepath dedicated to handling fake entities properly